### PR TITLE
qa: mds Enable multimds killpoint tests

### PR DESCRIPTION
**Pull request by lxbsz**
This version has been impoved a lot, including hadfailover_rank(),
to make it have the same logic with hadfailover().

And also keeps retry by sleeping 5 seconds every time instead of
hard code waiting 150 seconds to speed up the test. And also some
others small fixings.

Fixes: http://tracker.ceph.com/issues/17835
Signed-off-by: Sidharth Anupkrishnan <sanupkri@redhat.com>
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by lxbsz** [comment]
This is based @sidharthanup's previous PR https://github.com/ceph/ceph/pull/28004 and have some improvements about it.

File: qa/tasks/cephfs/filesystem.py [hunk 1]
```diff
@@ -207,6 +207,18 @@ def hadfailover(self, status):
         #all matching
         return False
 
+    def hadfailover_rank(self, fscid, status, rank):
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py:210]: This does an implicit `ceph fs dump` in `Filesystem.get_rank`.

I think a better place for this is in `FSStatus` with this signature:

`def had_failover_rank(self, fscid, rank, status2):`

File: qa/tasks/cephfs/test_exports.py [hunk 2]
```diff
@@ -517,3 +517,138 @@ def test_ephemeral_pin_shrink_mds(self):
...
+        result = self.fs.rank_asok(command, rank=0, status=status)
+        assert(result["success"])
+
+        command = ["config", "set", "mds_kill_import_at", str(importv)]
+        result = self.fs.rank_asok(command, rank=1, status=status)
+        assert(result["success"])
+
+        try:
+            # This should kill either or both MDS process
+            self.mount_a.setfattr("abc", "ceph.dir.pin", "1")
```
Comment (batrick) [inline qa/tasks/cephfs/test_exports.py:586]: What's preventing the balancer from doing part of this export before reaching this point? (I think we need to pin the directory to rank 0 before populating it.)

**Review by batrick — CHANGES_REQUESTED** [review]
@lxbsz please run through QA to verify it works when you're done adjusting the code.

Thanks a lot for picking this up.

File: qa/tasks/cephfs/filesystem.py (hunk 1 above)
Comment (lxbsz) [reply qa/tasks/cephfs/filesystem.py:210]: Looks good to me. Will update it.

File: qa/tasks/cephfs/test_exports.py (hunk 2 above)
Comment (lxbsz) [reply qa/tasks/cephfs/test_exports.py:586]: Good point. Yeah, actually we cannot be sure the `abc` dir is already pinned or auth in rank 0.

Comment (lxbsz) [reply qa/tasks/cephfs/test_exports.py:586]: > @lxbsz please run through QA to verify it works when you're done adjusting the code.
> 
Sure, will do that.

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Comment by lxbsz** [comment]
Today I test it more and found several bugs in MDS migrate code, I need more time to debug and fix it in late future.

**Comment by lxbsz** [comment]
jenkins test make check

**Comment by lxbsz** [comment]
jenkins test make check arm64

**Review by batrick — CHANGES_REQUESTED** [review]
I'm not sure `mds: set session state to _CLOSED when replaying the EImportStart` is correct. The MDSMap contains an `export_targets` set to get clients to connect to potential importers of subtrees.

**Comment by lxbsz** [comment]
> I'm not sure `mds: set session state to _CLOSED when replaying the EImportStart` is correct. The MDSMap contains an `export_targets` set to get clients to connect to potential importers of subtrees.

Since when writing journal the session was in _CLOSED state, after that in some killpoints the session maybe still in _CLOSED state, in some later it will be _OPENED, so just set it to _OPENED is not correct IMO. When replaying it just set it to the initial state, which is the same with when the journal is flushed.

I tried that weeks ago, it didn't work for me. If the above won't work for your I will try it again next week.

**Comment by lxbsz** [comment]
The teuthology tests passed, but this only for fuse mounter: https://pulpito.ceph.com/xiubli-2021-08-11_09:50:57-fs-wip-lxb-20210811-killpoint-1444-distro-basic-smithi/

For the kclient we still need one patch, I will post that later.

**Comment by lxbsz** [comment]
jenkins retest this please

**Comment by lxbsz** [comment]
@vshankar I think we still need this to test the exporting.

**Comment by vshankar** [comment]
> @vshankar I think we still need this to test the exporting.

I was not tracking this and I haven't gone through the changes. Any work that is pending?

**Comment by lxbsz** [comment]
> > @vshankar I think we still need this to test the exporting.
> 
> I was not tracking this and I haven't gone through the changes. Any work that is pending?

No work is pending.

**Pull request closed without merging**